### PR TITLE
Prebid core & PBS Adapter: debugging tools to intercept bid requests and mock their response

### PR DESCRIPTION
**Pull request by dgirardi**
## Type of change

- [x] Feature
- [x] Does this change affect user-facing APIs or examples documented on http://prebid.org?

## Description of change
This adds the ability to 'intercept' bid requests with a mock response. An intercepted bid does not go through the normal `buildRequests` -> network -> `interpretResponse` flow; a dummy response is instead directly returned to the auction as if it originated from `interpretResponse`.

This is mostly transparent to other parts of the stack (validation, events, etc), with the exception of user syncs - their interface requires the bidder's HTTP response, so they won't run if all bids in the request are intercepted (and thus no network activity happens).

## Other information

Addresses https://github.com/prebid/Prebid.js/issues/4461
Docs PR: https://github.com/prebid/prebid.github.io/pull/3575

```javascript
pbjs.setConfig({
  debugging: {
    enabled: true,
    intercept: [
     // intercept rules are evaluated in order, first to match wins
      {
        when: {
          //  intercept all  bids that have adUnitCode === 'test-div'
          adUnitCode: 'test-div',
          //  (any other property of the bid object can be tested here)
        },
        then: {
          // mock their response with sane defaults and `cpm: 10`
          cpm: 10
          // anything added here will appear in the response
        }
      },
      {
        // fancier match & replace options
        when: {
          adUnitCode: /div/,
          mediaTypes: (mt) => mt.banner != null,
        },
        then: {
          width: (bid) => bid.mediaTypes.banner.sizes[0][0],
          height: (bid) => bid.mediaTypes.banner.sizes[0][1],
        }
      },

    ]
  }
});

```

### CODE REVIEW

**Comment by ChrisHuie** [comment]
@Fawke @robertrmartinez can yu guys check out this pr when you get the chance? 🙏

**Review by Fawke — COMMENTED** [review]
Hi @dgirardi,

Thanks for adding this feature to the Debugging module. Your code changes mostly look good to me and  I've also tested the feature with different combinations of `ruleDefs` and it seems to be doing its job.

Do you think the keyword `intercept` in the config definition should be changed to `intercepts`?

The only major concern that I have is that as a side effect of this change, some existing Analytics Adapters which rely on Prebid Events for reporting will start misreporting. 
For example,  an adapter listening for the `bidResponse` event and in the case when debugging is turned on, it will report false/mock values to its server.

A naive solution could be that a AnalyticAdapter checks if debugging is turned on/off and depending on that, send the values to the Analytics server. I don't know if there is some other way which doesn't require the Analytics adapters to make this change.

**Review by FilipStamenkovic — COMMENTED** [review]
I don't think this is the best approach to solve this issue.  
Your solution requires adding `debugging.js` for each bid adapter, so debugging existing bidders won't work without additional code.  
Also, this adds additional code to each bidder which would increase final prebid.js bundle on publishers page.  

I think optional solution would be something that:
- won't inflate pbjs size when using in production
- can be used at once with all bidders

Can we add some kind of module that replaces existing functions in prebid core in order to replace bid requests and mock their responses? Similar to what we have in our tests.

File: src/debugging.js [hunk 1]
```diff
@@ -22,21 +31,29 @@ function addHooks(overrides) {
...
   removeHooks();
   addHooks(overrides);
   logMessage(`bidder overrides enabled${fromSession ? ' from session' : ''}`);
 }
 
 export function disableOverrides() {
+  bidInterceptor.updateConfig({});
   removeHooks();
   logMessage('bidder overrides disabled');
 }
```
Comment (Fawke) [inline src/debugging.js:59]: Does it make sense to clear the sessionStorage for `pbjs:debugging` here? While testing it locally, even though I had set debugging `enabled: false`, I had to manually close the tab to make this take effect.

**Comment by dgirardi** [comment]
@Fawke,

> The only major concern that I have is that as a side effect of this change, some existing Analytics Adapters which rely on Prebid Events for reporting will start misreporting.
> For example, an adapter listening for the `bidResponse` event and in the case when debugging is turned on, it will report false/mock values to its server.

I view that as a feature not a bug - a valid use case for this would be testing analytics adapters. Is there a reason analytics should be different from the rest? The mock values would only be reported when a publisher is explicitly doing testing.

**Comment by dgirardi** [comment]
@FilipStamenkovic,

> Your solution requires adding `debugging.js` for each bid adapter, so debugging existing bidders won't work without additional code.

Why? I have a special case for the PBS adapter (because that is itself a special case, not really an adapter) but nothing needs to be changed for anyone else (unless I am missing something).

The decision to make this core instead of a separate debugging module was made offline with @bretg, I believe to support the browser extension tooling. I am open to break this out into an optional module if the community agrees on that.

**Comment by bretg** [comment]
I think basic debugging support is worth having in PBJS-core.

The question of course is "what comprises basic debugging support"?

I'm often called on to debug publisher pages and find it difficult and time-consuming to attempt to build a custom debug-enabled PBJS package and fake the browser into loading it. 

So I think that a few hundred bytes of built-in debug support is worthwhile. But sounds like the community needs to have a debate about what specific debug features are worth the bytes and which need to be pushed to an optional module.

**Comment by Fawke** [comment]
> @Fawke,
> 
> > The only major concern that I have is that as a side effect of this change, some existing Analytics Adapters which rely on Prebid Events for reporting will start misreporting.
> > For example, an adapter listening for the `bidResponse` event and in the case when debugging is turned on, it will report false/mock values to its server.
> 
> I view that as a feature not a bug - a valid use case for this would be testing analytics adapters. Is there a reason analytics should be different from the rest? The mock values would only be reported when a publisher is explicitly doing testing.

I never said it is a bug but is a side effect of this change. I agree this would be helpful in testing Analytics adapters but we would need a way to differentiate bids originating from when debugging is turned on vs when it is off to avoid any discrepancies in reporting. @bretg had raised a similar point in this [comment](https://github.com/prebid/Prebid.js/issues/4461#issuecomment-564658039)

**Comment by FilipStamenkovic** [comment]
@dgirardi 
> Why? I have a special case for the PBS adapter (because that is itself a special case, not really an adapter) but nothing needs to be changed for anyone else (unless I am missing something).

My bad, I didn't test this properly. I just tested it with some random bidder and it works.

> Sounds like the community needs to have a debate about what specific debug features are worth the bytes and which need to be pushed to an optional module.

@bretg With these changes prebid-core.js is 9 KB heavier than on master branch, I'm not sure if this is a lot or not.

**Comment by dgirardi** [comment]
@Fawke  @bretg I need help understanding what exactly needs to be protected from "fake" bids; it seems odd that we'd care about polluting analytics but not for example the "bid won" handlers. I am not familiar with publisher workflows but I'd imagine they'd not be too interested in analytics for their own development activity - that is going to be an outlier regardless of whether the data stream is genuine or not. How does that work?

**Comment by dgirardi** [comment]
This was discussed offline today and we agreed to make it an optional module and have a strategy (in a separate PR) for dynamically loading it into production environments that may not have it.

**Comment by dgirardi** [comment]
Updated to have new debugging logic in an optional module. I have kept existing functionality (in `src/debugging.js`) mostly untouched except for some coordination to share `sessionStorage` logic. Eventually - when we have a way to load it asynchronously - I think it should all be moved into the debugging module, but I didn't want to remove it from core until that's ready.

**Comment by bretg** [comment]
> what exactly needs to be protected from "fake" bids; it seems odd that we'd care about polluting analytics

How about we add a "debug bid" onto any bid response object generated this way. Then analytics adapters can weed them out if they'd like

File: src/debugging.js (hunk 1 above)
Comment (dgirardi) [reply src/debugging.js]: I don't know if it has changed after my latest updates, but both session storage and debug hooks should be cleared with `setConfig({debugging: {enabled: false}})`. Do you still see this problem?

Comment (Fawke) [reply src/debugging.js]: Seems to be working now. Thanks.

**Review by Fawke — APPROVED** [review]

**Comment by Fawke** [comment]
> How about we add a "debug bid" flag onto any bid response object generated this way. Then analytics adapters can weed them out if they'd like

Just one thing to note, maybe I'm wrong, but I don't think this approach would catch/stop analytics adapters listening for events like, `auctionInit`, `bidRequested` and all those events which get triggered before we'd get a `bidResponse` event.

**Review by msm0504 — APPROVED** [review]

**Pull request merged**